### PR TITLE
Autocomplete 163

### PR DESCRIPTION
**Pull request by ebracho**
No css yet and I have a couple more ideas for features, but I think this is a good start.

Also, I created a new service called utility. It is a place that we can put useful functions we need that aren’t tied to a specific controller/directive/service/.

### CODE REVIEW

**Comment by tmjam** [comment]
@ebracho please update your branch with master

**Comment by ebracho** [comment]
@tmjam Done.

**Comment by tmjam** [comment]
@ebracho looks like there are more conflicts after the bug fix merge

**Comment by ebracho** [comment]
@tmjam Okay I think its all up to date now.

File: src/bossy.js [hunk 1]
```diff
@@ -13,12 +13,14 @@
...
+        'bossy.utility',
 		'bossy.calendar',
 		'bossy.data',
 		'bossy.form',
         'bossy.graph',
 		'bossy.input',
 		'bossy.schema',
-		'bossy.tooltip'
+		'bossy.tooltip',
+        'bossy.autocomplete'
```
Comment (radfahrer) [inline src/bossy.js]: Please fix the indentation so that all these directives are indented to the same level.

**Comment by ebracho** [comment]
@radfahrer Fixed. Is there any reason this file uses tabs instead of spaces like the rest of the repository?

**Comment by radfahrer** [comment]
@glawson tabs vs spaces?

File: src/directives/templates/bossy.autocomplete.html [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<div>
```
Comment (radfahrer) [inline src/directives/templates/bossy.autocomplete.html]: Is this file being used? It looks like the template is being defined inline in the directive. @glawson might know more.

File: src/services/bossy.bktree.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            return mat[str1.length - 1][str2.length - 1];
+        }
+
+        /**
+         * http://blog.notdot.net/2007/4/Damn-Cool-Algorithms-Part-1-BK-Trees
+         *
+         * @param String str
+         */
+        var Node = function(str) {
```
Comment (glawson) [inline src/services/bossy.bktree.js]: Please use another variable name, Node might be confusing.

File: src/directives/templates/bossy.autocomplete.html (hunk 2 above)
Comment (glawson) [inline src/directives/templates/bossy.autocomplete.html]: Good question, I'm not seeing any reference to it.

Comment (ebracho) [inline src/directives/templates/bossy.autocomplete.html]: Originally I wanted to load the template from this file, but I couldn't get templateUrl to work with preview. I'll remove the file.

**Comment by ebracho** [comment]
I thinking that it might make more sense to have the bktree and utility services inside the bossy.autocomplete module rather than in their own modules, to make things less spread out and more understandable. What do you guys think?

**Comment by glawson** [comment]
I agree, yes.

Geoff

Geoff Lawson
Front-End Software Architect
------------------------------
Build.com, Inc.
Smarter Home Improvement™
P: 800.375.3403

**Comment by ebracho** [comment]
I set up a demo for this directive. You can try it out here: http://ebracho.github.io/AutocompleteDemo/

Demo source: https://github.com/ebracho/AutocompleteDemo/tree/gh-pages

**Pull request merged**